## Data Exploration

In [1]:
# Step 1: Import necessary libraries
import pandas as pd

In [2]:
# Step 2: Load the dataset into a Pandas DataFrame
file_path = "/Users/deeptilalwani/Documents/Data Science/code clause/task 1/creditcard.csv"
data = pd.read_csv(file_path)


In [3]:
# Step 3: Explore the dataset
# 3.1: Display the first few rows of the dataset
print("First few rows of the dataset:")
print(data.head())

First few rows of the dataset:
   Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

 

In [4]:
# 3.2: Get information about the dataset
print("\nInformation about the dataset:")
print(data.info())


Information about the dataset:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 n

In [5]:
# 3.3: Get summary statistics of the dataset
print("\nSummary statistics of the dataset:")
print(data.describe())



Summary statistics of the dataset:
                Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  1.168375e-15  3.416908e-16 -1.379537e-15  2.074095e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   9.604066e-16  1.487313e-15 -5.556467e-16  1.213481e-16 -2.406331e-15   
std   

In [6]:
# 3.4: Check the distribution of the target variable (Class)
print("\nDistribution of the target variable (Class):")
print(data['Class'].value_counts())


Distribution of the target variable (Class):
0    284315
1       492
Name: Class, dtype: int64


## Data Preprocessing

In [7]:
# Check for missing values
print("\nMissing values in the dataset:")
print(data.isnull().sum())


Missing values in the dataset:
Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64


Since there are no missing values, we can skip the imputation step. We'll proceed with handling class imbalance using SMOTE.

In [8]:
#  Handling Class Imbalance
# Check the distribution of the target variable (Class)
print("\nDistribution of the target variable (Class):")
print(data['Class'].value_counts())


Distribution of the target variable (Class):
0    284315
1       492
Name: Class, dtype: int64


As we can see, there is a significant class imbalance in the dataset. The majority class (non-fraudulent transactions, labeled as 0) has 284,315 instances, while the minority class (fraudulent transactions, labeled as 1) has only 492 instances.

we'll use the Synthetic Minority Over-sampling Technique (SMOTE) to oversample the minority class. This technique generates synthetic samples for the minority class to balance the class distribution.

In [9]:
#  Handling Class Imbalance using SMOTE
from imblearn.over_sampling import SMOTE


In [10]:
# Separate features (X) and target variable (y)
X = data.drop('Class', axis=1)
y = data['Class']

In [11]:
# Initialize SMOTE
smote = SMOTE(random_state=42)

In [12]:
# Apply SMOTE to create synthetic samples for the minority class
X_resampled, y_resampled = smote.fit_resample(X, y)


In [13]:
# Check the distribution of the target variable after SMOTE
print("\nDistribution of the target variable (Class) after SMOTE:")
print(y_resampled.value_counts())


Distribution of the target variable (Class) after SMOTE:
0    284315
1    284315
Name: Class, dtype: int64


Now that the class distribution is balanced, we can proceed with applying an anomaly detection algorithm.

## Model Development

In [14]:
# Anomaly Detection using Isolation Forest
from sklearn.ensemble import IsolationForest


In [15]:
# Initialize the Isolation Forest model
isolation_forest = IsolationForest(contamination='auto', random_state=42)

In [16]:
# Fit the model to the data and predict anomalies
y_pred = isolation_forest.fit_predict(X_resampled)

In [17]:
# Anomalies are labeled as -1, and normal instances are labeled as 1
# Convert the labels to match the original dataset (0 for normal, 1 for anomaly)
y_pred[y_pred == 1] = 0  # Normal instances
y_pred[y_pred == -1] = 1  # Anomalies


Anomalies are predicted using Isolation Forest, and their labels are converted to match the original dataset. Finally, the distribution of predicted anomalies is printed.



In [18]:
# Check the distribution of predicted anomalies
print("\nDistribution of predicted anomalies:")
print(pd.Series(y_pred).value_counts())


Distribution of predicted anomalies:
0    495783
1     72847
dtype: int64


Anomalies are labeled as 1, and normal instances are labeled as 0.
According to the distribution, there are 72,847 instances predicted as anomalies (labeled as 1) and 495,783 instances predicted as normal (labeled as 0).


#### Let's experiment with different anomaly detection algorithms and hyperparameters to find the best-performing one. We'll use cross-validation to evaluate the models and select the one with the best performance.

In [19]:
from sklearn.model_selection import cross_val_score
from sklearn.neighbors import LocalOutlierFactor
from sklearn.svm import OneClassSVM
from sklearn.covariance import EllipticEnvelope


In [20]:
# Reduce the dataset size by using a smaller subset (e.g., first 10000 samples)
X_subset = X_resampled[:10000]
y_subset = y_resampled[:10000]


In [21]:
# Define a dictionary to store the models and their hyperparameters
models = {
    "Isolation Forest": IsolationForest(contamination='auto', random_state=42),
    "Local Outlier Factor": LocalOutlierFactor(novelty=True),
    "One-Class SVM": OneClassSVM(gamma='auto'),
    "Elliptic Envelope": EllipticEnvelope(contamination=0.01 , support_fraction=0.7)
}

## Model Evaluation

In [22]:
from sklearn.metrics import precision_score, recall_score, f1_score, average_precision_score, roc_auc_score
from sklearn.model_selection import cross_val_predict

In [23]:
# Perform model evaluation for each model
for name, model in models.items():
    # Perform cross-validation using 5-fold cross-validation
    y_pred = cross_val_predict(model, X_subset, y_subset, cv=5, n_jobs=-1)
    
    # Convert labels to match the original dataset (0 for normal, 1 for anomaly)
    y_pred[y_pred == 1] = 0  # Normal instances
    y_pred[y_pred == -1] = 1  # Anomalies
    
    try:
        # Calculate evaluation metrics
        precision = precision_score(y_subset, y_pred)
        recall = recall_score(y_subset, y_pred)
        f1 = f1_score(y_subset, y_pred)
        auc_pr = average_precision_score(y_subset, y_pred)  
    
        # Print evaluation metrics
        print(f"Model: {name}")
        print(f"Precision: {precision:.4f}, Recall: {recall:.4f}, F1-score: {f1:.4f}, AUC-PR: {auc_pr:.4f}")
    except Exception as e:
        print(f"Error occurred for model {name}: {e}")

Model: Isolation Forest
Precision: 0.0788, Recall: 1.0000, F1-score: 0.1462, AUC-PR: 0.0788


/Users/deeptilalwani/Downloads/anaconda3/lib/python3.10/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but LocalOutlierFactor was fitted with feature names
  warnings.warn(
/Users/deeptilalwani/Downloads/anaconda3/lib/python3.10/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but LocalOutlierFactor was fitted with feature names
  warnings.warn(
/Users/deeptilalwani/Downloads/anaconda3/lib/python3.10/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but LocalOutlierFactor was fitted with feature names
  warnings.warn(
/Users/deeptilalwani/Downloads/anaconda3/lib/python3.10/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but LocalOutlierFactor was fitted with feature names
  warnings.warn(
/Users/deeptilalwani/Downloads/anaconda3/lib/python3.10/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but LocalOutlie

Model: Local Outlier Factor
Precision: 0.0038, Recall: 1.0000, F1-score: 0.0076, AUC-PR: 0.0038
Model: One-Class SVM
Precision: 0.0038, Recall: 1.0000, F1-score: 0.0076, AUC-PR: 0.0038


/Users/deeptilalwani/Downloads/anaconda3/lib/python3.10/site-packages/sklearn/covariance/_robust_covariance.py:183: RuntimeWarning: Determinant has increased; this should not happen: log(det) > log(previous_det) (-117.578785769274148 > -118.408508434421947). You may want to try with a higher value of support_fraction (current value: 0.700).
  warnings.warn(
/Users/deeptilalwani/Downloads/anaconda3/lib/python3.10/site-packages/sklearn/covariance/_robust_covariance.py:183: RuntimeWarning: Determinant has increased; this should not happen: log(det) > log(previous_det) (-124.658461522104488 > -126.408286789320044). You may want to try with a higher value of support_fraction (current value: 0.700).
  warnings.warn(
/Users/deeptilalwani/Downloads/anaconda3/lib/python3.10/site-packages/sklearn/covariance/_robust_covariance.py:183: RuntimeWarning: Determinant has increased; this should not happen: log(det) > log(previous_det) (-120.308560132829996 > -121.676387675824259). You may want to try w

Model: Elliptic Envelope
Precision: 0.1154, Recall: 0.3158, F1-score: 0.1690, AUC-PR: 0.0390


/Users/deeptilalwani/Downloads/anaconda3/lib/python3.10/site-packages/sklearn/covariance/_robust_covariance.py:183: RuntimeWarning: Determinant has increased; this should not happen: log(det) > log(previous_det) (-119.455387305170717 > -122.657666482772086). You may want to try with a higher value of support_fraction (current value: 0.700).
  warnings.warn(


The performance metric used is accuracy

It seems like all models have very low precision, recall, F1-score, and AUC-PR values, indicating that they are not performing well for fraud detection on this dataset.

Low recall indicates that the models are failing to identify a significant portion of the actual fraudulent transactions (i.e., many fraudulent transactions are being missed). Low precision suggests that a large number of the transactions flagged as fraudulent are actually legitimate.

### By fine-tuning the parameters of our chosen model, we can potentially achieve better performance without significantly altering the dataset or feature engineering.

## hyperparameter tuning
##### hyperparameter tuning process using grid search with a Random Forest Classifier

Based on the evaluation metrics, Isolation Forest seems to have the highest recall among the models, indicating that it's better at detecting anomalies. However, the precision is low, which means it might also misclassify some normal instances as anomalies. On the other hand, Elliptic Envelope has higher precision but lower recall compared to Isolation Forest.

prioritizing detecting as many fraud cases as possible (even if it means some false positives), we might prefer Isolation Forest.

In [24]:
from sklearn.ensemble import IsolationForest
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import train_test_split

In [25]:
# Define the hyperparameter grid
param_grid = {
    'n_estimators': [50, 100, 200],
    'max_samples': ['auto', 0.5, 0.7, 0.9],
    'contamination': [0.01, 0.05, 0.1]
}


In [26]:
# Instantiate the Isolation Forest model
isolation_forest = IsolationForest(random_state=42)


In [27]:
# Split the original dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

In [28]:
# Sample a subset of your original dataset
X_subset, _, y_subset, _ = train_test_split(X_train, y_train, train_size=0.1, stratify=y_train, random_state=42)

In [29]:
import warnings

# Suppress warnings
warnings.filterwarnings("ignore", category=UserWarning)


In [30]:
# Perform grid search cross-validation on the subset
grid_search = GridSearchCV(isolation_forest, param_grid, cv=5, scoring='roc_auc')
grid_search.fit(X_subset, y_subset)

GridSearchCV(cv=5, estimator=IsolationForest(random_state=42),
             param_grid={'contamination': [0.01, 0.05, 0.1],
                         'max_samples': ['auto', 0.5, 0.7, 0.9],
                         'n_estimators': [50, 100, 200]},
             scoring='roc_auc')

In [31]:
# Get the best model
best_isolation_forest = grid_search.best_estimator_


In [32]:
# Make predictions on the full dataset
y_pred = best_isolation_forest.predict(X_test)

# Calculate evaluation metrics
precision = precision_score(y_test, y_pred, average='micro')
recall = recall_score(y_test, y_pred, average='micro')
f1 = f1_score(y_test, y_pred, average='micro')
roc_auc = roc_auc_score(y_test, y_pred)


In [33]:
# Print or display the evaluation metrics
print("Best Isolation Forest Model Evaluation on Full Dataset:")
print(f"Precision: {precision:.4f}, Recall: {recall:.4f}, F1-score: {f1:.4f}, ROC AUC: {roc_auc:.4f}")

Best Isolation Forest Model Evaluation on Full Dataset:
Precision: 0.0008, Recall: 0.0008, F1-score: 0.0008, ROC AUC: 0.2394


The precision, recall, and F1-score are all around 0.0008, indicating poor performance. Additionally, the ROC AUC score is 0.2394, which also suggests that the model is not performing well.


Feature Engineering: Explore additional feature engineering techniques to extract more relevant information from the data that could improve model performance.

### Feature Engineering

In [34]:
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
import numpy as np


In [35]:
# Define a function for feature engineering
def feature_engineering(data):
    # Separate features and target variable
    X = data.drop(columns=['Class'])
    y = data['Class']
    
    # Normalize/Scale features
    scaler = StandardScaler()
    X_scaled = scaler.fit_transform(X)
    
    # Dimensionality reduction with PCA
    pca = PCA(n_components=10)  # Adjust the number of components as needed
    X_pca = pca.fit_transform(X_scaled)
    
    # Concatenate original features with PCA-transformed features
    X_engineered = np.hstack((X_scaled, X_pca))
    
    return X_engineered, y


In [36]:
# Apply feature engineering to your dataset
X_engineered, y = feature_engineering(data)

Noe we train our anomaly detection mdoel on the engineered features

In [37]:
# Split the engineered features and target variable into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_engineered, y, test_size=0.2, random_state=42, stratify=y)

In [38]:
# Train your anomaly detection model (e.g., Isolation Forest) on the training data
# (Assuming you have already imported and initialized the Isolation Forest model)
isolation_forest = IsolationForest(contamination='auto', random_state=42)
isolation_forest.fit(X_train)


IsolationForest(random_state=42)

In [39]:
# Make predictions on the testing data
y_pred = isolation_forest.predict(X_test)


In [40]:
# Calculate evaluation metrics
precision = precision_score(y_test, y_pred, average='weighted')
recall = recall_score(y_test, y_pred, average='weighted')
f1 = f1_score(y_test, y_pred, average='weighted')
roc_auc = roc_auc_score(y_test, y_pred)

In [41]:
# Print the evaluation metrics
print("Model Evaluation Metrics:")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1-score: {f1:.4f}")
print(f"ROC AUC: {roc_auc:.4f}")


Model Evaluation Metrics:
Precision: 0.0000
Recall: 0.0002
F1-score: 0.0000
ROC AUC: 0.0914


The evaluation metrics indicate very low performance for the model. The precision, recall, and F1-score are close to zero, and the ROC AUC is also very low. This suggests that the model is not effectively capturing the true positive instances of fraud.



using other basic classification algorithms:

In [42]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score


In [43]:
# Initialize the classifiers
lr = LogisticRegression()
svc = SVC()
rf = RandomForestClassifier()


In [44]:
# Create a dictionary of classifiers
models = {'Logistic Regression': lr, 'SVM': svc, 'Random Forest': rf}

In [45]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [46]:
# Train and evaluate each model
for name, model in models.items():
    # Train the model
    model.fit(X_train, y_train)
    
    # Make predictions
    y_pred = model.predict(X_test)
    
    # Calculate evaluation metrics
    accuracy = accuracy_score(y_test, y_pred)
    precision = precision_score(y_test, y_pred)
    recall = recall_score(y_test, y_pred)
    f1 = f1_score(y_test, y_pred)
    roc_auc = roc_auc_score(y_test, y_pred)
    
    # Print the evaluation metrics
    print(f'Model: {name}')
    print(f'Accuracy: {accuracy:.4f}, Precision: {precision:.4f}, Recall: {recall:.4f}, F1-score: {f1:.4f}, ROC AUC: {roc_auc:.4f}')

Model: Logistic Regression
Accuracy: 0.9986, Precision: 0.6111, Recall: 0.5612, F1-score: 0.5851, ROC AUC: 0.7803
Model: SVM
Accuracy: 0.9983, Precision: 0.0000, Recall: 0.0000, F1-score: 0.0000, ROC AUC: 0.5000
Model: Random Forest
Accuracy: 0.9996, Precision: 0.9747, Recall: 0.7857, F1-score: 0.8701, ROC AUC: 0.8928


Logistic Regression achieved high accuracy (99.86%) and reasonably good precision (61.11%), recall (56.12%), F1-score (58.51%), and ROC AUC (78.03%).
SVM achieved high accuracy (99.83%) but had very low precision, recall, and F1-score, indicating that it may not be suitable for this task.
Random Forest achieved the highest accuracy (99.96%) among the three models and had excellent precision (97.40%), recall (76.53%), F1-score (85.71%), and ROC AUC (88.26%).

Considering these results, Random Forest seems to be the best-performing model for this classification task, as it achieved the highest accuracy and good values for precision, recall, F1-score, and ROC AUC.